### ISSUE TITLE
Post-merge issues of decoupling the interface

### ISSUE THREAD

**Issue by vitaly-t** [vitaly-t/iter-ops#203]
This is to address issues after PR #202, as a major rework of the architecture.

@RebeccaStevens Let's address any and all issues of that merge here, rather than the PR itself.

**Comment by vitaly-t** [comment]
I really wish I merged that big change into a dev branch, shouldn't code when falling asleep :smile:

I needed a solid version, with the #200 change, so here what I did today:

* Cleaned up the list of branches, by deleting all the obsolete crap and merged ones. I even deleted `develop` branch.
* Created new branch `dev` from my last commit for #200, just before merging the decouple PR.
* Released [v2.7.0](https://github.com/vitaly-t/iter-ops/releases/tag/2.7.0) with all the latest stuff (prior decoupling) from `dev` branch.

So we now have `main` and `dev` with what should be swapped in them, but it will have to do for the time being, alas.

We still have that `2.7.0-beta.0` with all the merged decoupling, which I've been using to test out the changes, not successfully for now, that's why we have this issue, so let's address it here ;)

**Comment by RebeccaStevens** [comment]
I've updated dev to have the new changes.
If you run this, it should get main back to where you want it.

```
git checkout main
git reset --hard 2.7.0
git push --force
```

**Comment by vitaly-t** [comment]
Just tried it, `git reset --hard 2.7.0`  produces:

> fatal: ambiguous argument '2.7.0': unknown revision or path not in the working tree.

**Comment by RebeccaStevens** [comment]
Does 2.7.0 show up when you run `git tag --list`?

**Comment by vitaly-t** [comment]
It does not, 2.6.1 comes up as the latest, which is odd, because I can see 2.7.0 in the list of tags on the website...

![image](https://user-images.githubusercontent.com/5108906/204247893-adf40b95-bbd9-4b4d-b4a4-7884ebf03155.png)

**Comment by RebeccaStevens** [comment]
Try

`git fetch --all --tags`

**Comment by vitaly-t** [comment]
That did it,...will release the direct commit hook now, and try again :)

**Comment by RebeccaStevens** [comment]
With regard to the documentation, there seems to be a bunch of "references" for types that are exported by multiple namespaces. It seems the `async` namespace is getting them as first-class entries while the `index` and `sync` namespaces just reference the `async` ones. I don't know if there's a way to make is so the `index` namespace is the one that gets them as first-class entries or not.

One simple solution could be just to only export those types in the `index` namespace rather than in all namespaces.

**Comment by vitaly-t** [comment]
Branch re-basement finished, looks good, cheers!

**Comment by RebeccaStevens** [comment]
This code should work in with `2.7.0-beta.0`.

```
import { pipe, map } from "iter-ops/sync";

const input = [1, 2, 3];
const output = pipe(
  input,
  map((x) => x * 2)
);

console.log(output);
```

Note: If you are using TypeScript, you'll need to set `moduleResolution` to either `Node16` or `NodeNext` to get the types to resolve properly.

**Comment by vitaly-t** [comment]
> you'll need to set moduleResolution to either Node16 or NodeNext to get the types to resolve properly.

So how is it going to work inside web browsers?

**Comment by RebeccaStevens** [comment]
In what regard? web browsers don't need type information.

**Comment by vitaly-t** [comment]
Angular, for example sets `"moduleResolution": "node",` If I change it, resolution becomes broken for all Angular modules.

**Comment by vitaly-t** [comment]
Also, if the library requires tweaking `tsconfig` manually after installation, it'll likely drive developers away.

**Comment by RebeccaStevens** [comment]
So `"moduleResolution": "Node16"` is the newer way of doing things.

[AngularJS support has officially ended as of January 2022](https://github.com/angular/angular.js#:~:text=AngularJS%20support%20has%20officially%20ended%20as%20of%20January%202022) so AngularJS might not ever support `"moduleResolution": "Node16"`.

In this case, you'd need to do things to other (older) way.

```ts
import { sync } from "iter-ops";
const { pipe, map } = sync;

const input = [1, 2, 3];
const output = pipe(
  input,
  map((x) => x * 2)
);

console.log(output);
```

**Comment by vitaly-t** [comment]
I wasn't talking about the ancient AngularJS, but the modern Angular v15, which is a totally different thing these days :)

b.t.w. That's my primary tool for Web. I'm an Angular specialist ;)

**Comment by RebeccaStevens** [comment]
So that code example in my previous comment should work for you for now until that linked issue is resolved.

**Comment by vitaly-t** [comment]
I just redid my tests against the very latest Angular, and it behaves differently again...

Now setting `"moduleResolution": "NodeNext",` does nothing.

And for some reasons it wants to find things inside `iter-ops/dist/async`, as shown below, from which pipe types are casted incorrectly...

![image](https://user-images.githubusercontent.com/5108906/204260117-e44a04a0-0103-4df0-a6a5-07fb145f0190.png)

**Comment by RebeccaStevens** [comment]
Here's the way it should be able to be imported:

Option 0 - No changes:

`"moduleResolution": "Node"`
```ts
import { pipe, delay, tap } from "iter-ops";
```

Option 1 - Classic resolution to access the new namespaces:

`"moduleResolution": "Node"`
```ts
import { async } from "iter-ops";
const { pipe, delay, tap } =  async;
```

Option 2 - Modern resolution to access the new namespaces:

`"moduleResolution": "Node16"`
```ts
import { pipe, delay, tap } from "iter-ops/async";
```


Option 3 - Access the internals to access the new namespaces:

`"moduleResolution": "Node"`
```ts
import { pipe, delay, tap } from "iter-ops/dist/async";
```

**Comment by vitaly-t** [comment]
I'm not sure how we are going to proceed here, I'm just gonna leave it in your hands, for now. Version 2.7.0 installs and works without any issues everywhere, and we should have the same expectation for this branch, which is currently not the case, as far as I can see it...

For the time being, I might just focus on something simpler... like picking some operator from [custom operators](https://github.com/vitaly-t/iter-ops-extras), and adding it to the library. Operators like `appendIf`, `prependIf` + `replaceIf` are probably the most valuable there on the list.

**Comment by RebeccaStevens** [comment]
You should be able to use the library exactly the same as you could before.
The new changes are an optional alternative way to do things.

**Comment by RebeccaStevens** [comment]
I just updated the tests and fixed up a few bugs that were revealed when doing so.
Do you want to release a new beta? `3.0.0-beta.0`

**Comment by vitaly-t** [comment]
What are the breaking changes? Trying to figure out if 2.8.0-beta.0 would be more appropriate...

**Comment by RebeccaStevens** [comment]
They are very minor ones. I might even be able to remove them and make it so there aren't any.

The type `UnknownIterableOrIterator` no longer exists.

Other than that, there may not actually be any. The other major changes I made don't actually seem to be publicly visible. `2.8.0-beta.0` might be the better choice.

**Comment by vitaly-t** [comment]
`2.8.0-beta.0` published ;)

**Comment by vitaly-t** [comment]
I took some time going over the `dev` branch, and the custom operators PR, plus tried to play with it locally...

I honestly do see this as an improvement. It might offer better type casting (I'm guessing here), but it replaces the plain `operator` syntax with `sync.operator` and `async.operator` syntax, and splits a single-file implementation into 3-4 files, with essentially repeated implementation. All that, combined with such a radical change in terms of the structure... I'll be honest, I'm not warming up to this idea.

I like the current architecture, it is way simpler, the way I see it. And I most definitely wouldn't want to complicate the idea of simple custom operators, with one where you have to create 4 files for it... that's just gonna be a turn-off for most developers, IMO.

I really do not want to seem ungrateful for how much code you have written for it, it is an incredible job, but I'm not sure I want this change. I wish we discussed it first, or you just did a POC instead... Really sorry!

**Comment by RebeccaStevens** [comment]
I've made some changes to the dev branch to restructure things and simplify it a bit.

I think I might have also been able to solve the resolution issue so that `"moduleResolution": "Node16"` is no longer needed (untested - but I think it should work).

There are some type issues that have arisen from the rebase, it should be an easy fix.

Let me know what you think of this change.

**Comment by vitaly-t** [comment]
I have postponed this for now, and released [v3.0.0](https://github.com/vitaly-t/iter-ops/releases/tag/3.0.0) without it, for there were too many changes since v2.0.0 as it is.

**Comment by RebeccaStevens** [comment]
I believe I've now fixed up all the issues.
Please re-review when you have time.

**Comment by vitaly-t** [comment]
Published `4.0.0-beta.0` from the `dev` branch, to make evaluation easier.
